In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

### 1. Load the data. You may drop id or just ignore it in the rest of your analysis. You may also convert diagnosis into a numerical variable.

In [2]:
wbc = pd.read_csv('./wdbc.csv.bz2')
wbc.head(10)

id diagnosis  radius.mean  texture.mean  perimeter.mean  area.mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   
5    843786         M        12.45         15.70           82.57      477.1   
6    844359         M        18.25         19.98          119.60     1040.0   
7  84458202         M        13.71         20.83           90.20      577.9   
8    844981         M        13.00         21.82           87.50      519.8   
9  84501001         M        12.46         24.04           83.97      475.9   

   smoothness.mean  compactness.mean  concavity.mean  concpoints.mean  ...  \
0          0.11840           0.27760         0.30010          0.14710  ...   
1          0.08474           0.07864         0.08690          0.07017  ...   
2          0.10960           0.15990         0.19740          0.12790  ...   
3          0.14250           0.28390         0.24140          0.10520  ...   
4          0.10030           0.13280         0.19800          0.10430  ...   
5          0.12780           0.17000         0.15780          0.08089  ...   
6          0.09463           0.10900         0.11270          0.07400  ...   
7          0.11890           0.16450         0.09366          0.05985  ...   
8          0.12730           0.19320         0.18590          0.09353  ...   
9          0.11860           0.23960         0.22730          0.08543  ...   

   radius.worst  texture.worst  perimeter.worst  area.worst  smoothness.worst  \
0         25.38          17.33           184.60      2019.0            0.1622   
1         24.99          23.41           158.80      1956.0            0.1238   
2         23.57          25.53           152.50      1709.0            0.1444   
3         14.91          26.50            98.87       567.7            0.2098   
4         22.54          16.67           152.20      1575.0            0.1374   
5         15.47          23.75           103.40       741.6            0.1791   
6         22.88          27.66           153.20      1606.0            0.1442   
7         17.06          28.14           110.60       897.0            0.1654   
8         15.49          30.73           106.20       739.3            0.1703   
9         15.09          40.68            97.65       711.4            0.1853   

   compactness.worst  concavity.worst  concpoints.worst  symmetry.worst  \
0             0.6656           0.7119            0.2654          0.4601   
1             0.1866           0.2416            0.1860          0.2750   
2             0.4245           0.4504            0.2430          0.3613   
3             0.8663           0.6869            0.2575          0.6638   
4             0.2050           0.4000            0.1625          0.2364   
5             0.5249           0.5355            0.1741          0.3985   
6             0.2576           0.3784            0.1932          0.3063   
7             0.3682           0.2678            0.1556          0.3196   
8             0.5401           0.5390            0.2060          0.4378   
9             1.0580           1.1050            0.2210          0.4366   

   fracdim.worst  
0        0.11890  
1        0.08902  
2        0.08758  
3        0.17300  
4        0.07678  
5        0.12440  
6        0.08368  
7        0.11510  
8        0.10720  
9        0.20750  

[10 rows x 32 columns]

In [3]:
wbc.diagnosis = (wbc.diagnosis == 'M').astype(int)
wbc.head(10)

id  diagnosis  radius.mean  texture.mean  perimeter.mean  area.mean  \
0    842302          1        17.99         10.38          122.80     1001.0   
1    842517          1        20.57         17.77          132.90     1326.0   
2  84300903          1        19.69         21.25          130.00     1203.0   
3  84348301          1        11.42         20.38           77.58      386.1   
4  84358402          1        20.29         14.34          135.10     1297.0   
5    843786          1        12.45         15.70           82.57      477.1   
6    844359          1        18.25         19.98          119.60     1040.0   
7  84458202          1        13.71         20.83           90.20      577.9   
8    844981          1        13.00         21.82           87.50      519.8   
9  84501001          1        12.46         24.04           83.97      475.9   

   smoothness.mean  compactness.mean  concavity.mean  concpoints.mean  ...  \
0          0.11840           0.27760         0.30010          0.14710  ...   
1          0.08474           0.07864         0.08690          0.07017  ...   
2          0.10960           0.15990         0.19740          0.12790  ...   
3          0.14250           0.28390         0.24140          0.10520  ...   
4          0.10030           0.13280         0.19800          0.10430  ...   
5          0.12780           0.17000         0.15780          0.08089  ...   
6          0.09463           0.10900         0.11270          0.07400  ...   
7          0.11890           0.16450         0.09366          0.05985  ...   
8          0.12730           0.19320         0.18590          0.09353  ...   
9          0.11860           0.23960         0.22730          0.08543  ...   

   radius.worst  texture.worst  perimeter.worst  area.worst  smoothness.worst  \
0         25.38          17.33           184.60      2019.0            0.1622   
1         24.99          23.41           158.80      1956.0            0.1238   
2         23.57          25.53           152.50      1709.0            0.1444   
3         14.91          26.50            98.87       567.7            0.2098   
4         22.54          16.67           152.20      1575.0            0.1374   
5         15.47          23.75           103.40       741.6            0.1791   
6         22.88          27.66           153.20      1606.0            0.1442   
7         17.06          28.14           110.60       897.0            0.1654   
8         15.49          30.73           106.20       739.3            0.1703   
9         15.09          40.68            97.65       711.4            0.1853   

   compactness.worst  concavity.worst  concpoints.worst  symmetry.worst  \
0             0.6656           0.7119            0.2654          0.4601   
1             0.1866           0.2416            0.1860          0.2750   
2             0.4245           0.4504            0.2430          0.3613   
3             0.8663           0.6869            0.2575          0.6638   
4             0.2050           0.4000            0.1625          0.2364   
5             0.5249           0.5355            0.1741          0.3985   
6             0.2576           0.3784            0.1932          0.3063   
7             0.3682           0.2678            0.1556          0.3196   
8             0.5401           0.5390            0.2060          0.4378   
9             1.0580           1.1050            0.2210          0.4366   

   fracdim.worst  
0        0.11890  
1        0.08902  
2        0.08758  
3        0.17300  
4        0.07678  
5        0.12440  
6        0.08368  
7        0.11510  
8        0.10720  
9        0.20750  

[10 rows x 32 columns]

### 2. Create a summary table where you show means, ranges, and number of missings for each variable. In addition, add (Pearson) correlation between the diagnosis and the corresponding feature. You may include more statistics you consider useful. Order this information by the absolute value of correlation, so the table might look like

In [4]:
summary = pd.DataFrame({'variable' : [], 'correlation' : [], 'mean' : [], 'min' : [], 'max' : [], 'std' : [], 'missings' : []})
summary.variable = wbc.columns

for i in summary.variable:
    index = summary.index[summary['variable'] == i]
    summary.loc[index] = [i, wbc['diagnosis'].corr(wbc[i], method='pearson'), np.mean(wbc[i]), np.min(wbc[i]), np.max(wbc[i]), np.std(wbc[i]), wbc[i].isnull().sum()]

In [5]:
summary

variable  correlation          mean          min           max  \
0                  id     0.039769  3.037183e+07  8670.000000  9.113205e+08   
1           diagnosis     1.000000  3.725835e-01     0.000000  1.000000e+00   
2         radius.mean     0.730029  1.412729e+01     6.981000  2.811000e+01   
3        texture.mean     0.415185  1.928965e+01     9.710000  3.928000e+01   
4      perimeter.mean     0.742636  9.196903e+01    43.790000  1.885000e+02   
5           area.mean     0.708984  6.548891e+02   143.500000  2.501000e+03   
6     smoothness.mean     0.358560  9.636028e-02     0.052630  1.634000e-01   
7    compactness.mean     0.596534  1.043410e-01     0.019380  3.454000e-01   
8      concavity.mean     0.696360  8.879932e-02     0.000000  4.268000e-01   
9     concpoints.mean     0.776614  4.891915e-02     0.000000  2.012000e-01   
10      symmetry.mean     0.330499  1.811619e-01     0.106000  3.040000e-01   
11       fracdim.mean    -0.012838  6.279761e-02     0.049960  9.744000e-02   
12          radius.se     0.567134  4.051721e-01     0.111500  2.873000e+00   
13         texture.se    -0.008303  1.216853e+00     0.360200  4.885000e+00   
14       perimeter.se     0.556141  2.866059e+00     0.757000  2.198000e+01   
15            area.se     0.548236  4.033708e+01     6.802000  5.422000e+02   
16      smoothness.se    -0.067016  7.040979e-03     0.001713  3.113000e-02   
17     compactness.se     0.292999  2.547814e-02     0.002252  1.354000e-01   
18       concavity.se     0.253730  3.189372e-02     0.000000  3.960000e-01   
19      concpoints.se     0.408042  1.179614e-02     0.000000  5.279000e-02   
20        symmetry.se    -0.006522  2.054230e-02     0.007882  7.895000e-02   
21         fracdim.se     0.077972  3.794904e-03     0.000895  2.984000e-02   
22       radius.worst     0.776454  1.626919e+01     7.930000  3.604000e+01   
23      texture.worst     0.456903  2.567722e+01    12.020000  4.954000e+01   
24    perimeter.worst     0.782914  1.072612e+02    50.410000  2.512000e+02   
25         area.worst     0.733825  8.805831e+02   185.200000  4.254000e+03   
26   smoothness.worst     0.421465  1.323686e-01     0.071170  2.226000e-01   
27  compactness.worst     0.590998  2.542650e-01     0.027290  1.058000e+00   
28    concavity.worst     0.659610  2.721885e-01     0.000000  1.252000e+00   
29   concpoints.worst     0.793566  1.146062e-01     0.000000  2.910000e-01   
30     symmetry.worst     0.416294  2.900756e-01     0.156500  6.638000e-01   
31      fracdim.worst     0.323872  8.394582e-02     0.055040  2.075000e-01   

             std  missings  
0   1.249107e+08       0.0  
1   4.834925e-01       0.0  
2   3.520951e+00       0.0  
3   4.297255e+00       0.0  
4   2.427762e+01       0.0  
5   3.516048e+02       0.0  
6   1.405176e-02       0.0  
7   5.276633e-02       0.0  
8   7.964973e-02       0.0  
9   3.876873e-02       0.0  
10  2.739018e-02       0.0  
11  7.054156e-03       0.0  
12  2.770689e-01       0.0  
13  5.511634e-01       0.0  
14  2.020077e+00       0.0  
15  4.545101e+01       0.0  
16  2.999878e-03       0.0  
17  1.789244e-02       0.0  
18  3.015952e-02       0.0  
19  6.164861e-03       0.0  
20  8.259104e-03       0.0  
21  2.643745e-03       0.0  
22  4.828993e+00       0.0  
23  6.140854e+00       0.0  
24  3.357300e+01       0.0  
25  5.688565e+02       0.0  
26  2.281236e-02       0.0  
27  1.571982e-01       0.0  
28  2.084409e-01       0.0  
29  6.567455e-02       0.0  
30  6.181308e-02       0.0  
31  1.804539e-02       0.0

### 2.1 let's start with a logistic regression model where you explain diagnosis with concpoints.mean and fracdim.mean. Extract these data and split the result into training and validation chunks. Here are a few lines of example code that do something along these lines:

In [6]:
X = wbc[['concpoints.mean', 'fracdim.mean']].values
y = wbc.diagnosis
Xtrain, Xvalid, ytrain, yvalid = train_test_split(X, y, test_size = 0.2)

### 2. Now fit the model on training data

In [7]:
m = LogisticRegression(solver='lbfgs', C=1e9).fit(Xtrain, ytrain)

### 3. Predict the outcome on training data, and create a confusion matrix of the result

In [8]:
yhat = m.predict(Xtrain)
rmse = np.sqrt(np.mean((ytrain - yhat)**2))
rmse

0.26519741765271837

In [9]:
pd.crosstab(ytrain, yhat)

col_0        0    1
diagnosis          
0          273   15
1           17  150

### 4. Based on the confusion matrix, calculate accuracy and F-score

In [14]:
a = (150 + 273) / 455
a

0.9296703296703297

The accuracy is approximately 93%

In [15]:
p = 150 / (150 + 15)
r = 150 / (150 + 17)
f = 2 / (1 / p + 1 / r)
f

0.9036144578313253

f-score is 0.903

### 5. As we explained in class, it is not a good idea to compute the goodness of model on training data. So now predict the outcome on validation data, and create a confusion matrix of the result.

In [16]:
yhat = m.predict(Xvalid)
rmse = np.sqrt(np.mean((yvalid - yhat)**2))
rmse

0.3627381250550058

In [13]:
pd.crosstab(yvalid, yhat)

col_0       0   1
diagnosis        
0          62   7
1           8  37

### 6. Based on the confusion matrix, calculate accuracy and F-score.

In [17]:
acc = (37 + 62) / 114
acc

0.868421052631579

Accuracy is 86.8%

In [18]:
r = 37 / (37 + 8)
p = 37 / (37 + 7)
f = 2 / (1 / p + 1 / r)
f

0.8314606741573034

The F-score is 0.831

### 7. Comment the results: did the model perform better on training or testing data? Does it hint about overfitting?

The model performed better on training data as can be seen from the accuracy and a lower RMSE value. The difference is not a lot, hence suggesting that the model is not overfitting the data.

### 8. Repeat the above with a few other models: select different features or more features (you may select all of them too) and put into the model. Predict, display the confusion matrix and compute A and F on both trainig and validation dataset.

In [20]:
X0 = wbc.copy().drop(['diagnosis', 'id'], axis=1)
Xtrain, Xvalid, ytrain, yvalid = train_test_split(X0, y, test_size = 0.2)
m = LogisticRegression(solver='lbfgs', C=7e9).fit(Xtrain, ytrain)
m.coef_

C:\Users\divit\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


array([[-1.47410302, -0.58395765, -0.1083576 ,  0.00254228,  0.06412689,
         0.26804396,  0.37667919,  0.16358816,  0.06583769,  0.01684666,
        -0.06175471, -0.40037401, -0.23820026,  0.0875971 ,  0.00845151,
         0.05606118,  0.08553312,  0.02327079,  0.01447349,  0.00639245,
        -1.5585327 ,  0.61684846,  0.25137645,  0.02473055,  0.11330539,
         0.77524314,  0.98487825,  0.30815641,  0.1912903 ,  0.07723759]])

This warning can be ignored. It tells us that the logistic model and the data do not play well together but the problem is harmless here (too good a fit).

In [21]:
yhat = m.predict(Xtrain)
np.unique(yhat)

array([0, 1])

In [22]:
rmse = np.sqrt(np.mean((ytrain - yhat)**2))
rmse

0.22966770070528583

In [23]:
pd.crosstab(ytrain, yhat)

col_0        0    1
diagnosis          
0          278    9
1           15  153

In [24]:
acc = (153 + 278) / 455
acc

0.9472527472527472

The accuracy is approx 94.7%

In [26]:
r = 153 / (153 + 15)
p = 153 / (153 + 9)
F = 2 / (1 / p + 1 / r)
F

0.9272727272727271

F-score is 0.927272

In [27]:
yhat = m.predict(Xvalid)
rmse = np.sqrt(np.mean((yvalid - yhat)**2))
rmse

0.22941573387056177

In [28]:
pd.crosstab(yvalid, yhat)

col_0       0   1
diagnosis        
0          70   0
1           6  38

In [29]:
acc = (38 + 70) / 114
acc

0.9473684210526315

Accuracy of the model is 94.7%

In [30]:
r = 38 / (38 + 6)
p = 38 / (38 + 0)
f = 2 / (1 / p + 1 / r)
f

0.9268292682926829

f-score of the model is 0.926

### 9. Comment the results. Which model performed best in terms of A, F? Where the results on training data systematically better than on testing data?


The results were almost the same as both models had almost the same f-score, rmse accuracy.